File: maven-projects/info/src/test/java/org/apache/graphar/info/PropertyTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.graphar.info;

import org.apache.graphar.info.type.DataType;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

public class PropertyTest {

    private Property idProperty;
    private Property nameProperty;
    private Property optionalProperty;

    @Before
    public void setUp() {
        idProperty = TestDataFactory.createIdProperty();
        nameProperty = TestDataFactory.createProperty("name", DataType.STRING, false, true);
        optionalProperty = TestDataFactory.createProperty("optional", DataType.STRING, false, true);
    }

    @Test
    public void testPropertyBasicConstruction() {
        Property prop = TestDataFactory.createProperty("id", DataType.INT32, true, false);

        TestVerificationUtils.verifyProperty(prop, "id", true, false);
        Assert.assertEquals(DataType.INT32, prop.getDataType());
    }

    @Test
    public void testIdProperty() {
        TestVerificationUtils.verifyProperty(idProperty, "id", true, false);
        Assert.assertEquals(DataType.INT64, idProperty.getDataType());
    }

    @Test
    public void testPropertyWithStringType() {
        TestVerificationUtils.verifyProperty(nameProperty, "name", false, true);
        Assert.assertEquals(DataType.STRING, nameProperty.getDataType());
    }

    @Test
    public void testPropertyWithAllDataTypes() {
        Property boolProp = TestDataFactory.createProperty("flag", DataType.BOOL, false, false);
        Assert.assertEquals(DataType.BOOL, boolProp.getDataType());
        TestVerificationUtils.verifyProperty(boolProp, "flag", false, false);

        Property int32Prop = TestDataFactory.createProperty("count", DataType.INT32, true, false);
        Assert.assertEquals(DataType.INT32, int32Prop.getDataType());
        TestVerificationUtils.verifyProperty(int32Prop, "count", true, false);

        Property int64Prop =
                TestDataFactory.createProperty("bigCount", DataType.INT64, false, false);
        Assert.assertEquals(DataType.INT64, int64Prop.getDataType());
        TestVerificationUtils.verifyProperty(int64Prop, "bigCount", false, false);

        Property floatProp = TestDataFactory.createProperty("score", DataType.FLOAT, false, true);
        Assert.assertEquals(DataType.FLOAT, floatProp.getDataType());
        TestVerificationUtils.verifyProperty(floatProp, "score", false, true);

        Property doubleProp =
                TestDataFactory.createProperty("precision", DataType.DOUBLE, false, true);
        Assert.assertEquals(DataType.DOUBLE, doubleProp.getDataType());
        TestVerificationUtils.verifyProperty(doubleProp, "precision", false, true);

        Property stringProp = TestDataFactory.createProperty("text", DataType.STRING, false, true);
        Assert.assertEquals(DataType.STRING, stringProp.getDataType());
        TestVerificationUtils.verifyProperty(stringProp, "text", false, true);

        Property listProp = TestDataFactory.createProperty("items", DataType.LIST, false, true);
        Assert.assertEquals(DataType.LIST, listProp.getDataType());
        TestVerificationUtils.verifyProperty(listProp, "items", false, true);

        // Newly added data types
        Property dateProp =
                TestDataFactory.createProperty("creationDate", DataType.DATE, false, true);
        Assert.assertEquals(DataType.DATE, dateProp.getDataType());
        TestVerificationUtils.verifyProperty(dateProp, "creationDate", false, true);

        Property timestampProp =
                TestDataFactory.createProperty("createdAt", DataType.TIMESTAMP, false, true);
        Assert.assertEquals(DataType.TIMESTAMP, timestampProp.getDataType());
        TestVerificationUtils.verifyProperty(timestampProp, "createdAt", false, true);
    }

    @Test
    public void testPropertyFlagCombinations() {
        // Primary and non-nullable (typical ID field)
        Property primaryNonNull = TestDataFactory.createProperty("id", DataType.INT64, true, false);
        TestVerificationUtils.verifyProperty(primaryNonNull, "id", true, false);

        // Primary and nullable (unusual but valid)
        Property primaryNull =
                TestDataFactory.createProperty("optionalId", DataType.INT64, true, true);
        TestVerificationUtils.verifyProperty(primaryNull, "optionalId", true, true);

        // Non-primary and non-nullable (required field)
        Property requiredField =
                TestDataFactory.createProperty("required", DataType.STRING, false, false);
        TestVerificationUtils.verifyProperty(requiredField, "required", false, false);

        // Non-primary and nullable (optional field)
        TestVerificationUtils.verifyProperty(optionalProperty, "optional", false, true);
    }

    @Test
    public void testPropertyWithEmptyName() {
        Property emptyNameProp = TestDataFactory.createProperty("", DataType.STRING, false, true);
        TestVerificationUtils.verifyProperty(emptyNameProp, "", false, true);
        Assert.assertEquals(DataType.STRING, emptyNameProp.getDataType());
    }

    @Test
    public void testPropertyWithNullName() {
        Property nullNameProp = TestDataFactory.createProperty(null, DataType.STRING, false, true);
        Assert.assertNull(nullNameProp.getName());
        Assert.assertEquals(DataType.STRING, nullNameProp.getDataType());
        Assert.assertFalse(nullNameProp.isPrimary());
        Assert.assertTrue(nullNameProp.isNullable());
    }

    @Test
    public void testPropertyWithNullDataType() {
        Property nullTypeProp = TestDataFactory.createProperty("test", null, false, true);
        TestVerificationUtils.verifyProperty(nullTypeProp, "test", false, true);
        Assert.assertNull(nullTypeProp.getDataType());
    }

    @Test
    public void testPropertyEquality() {
        Property prop1 = TestDataFactory.createProperty("id", DataType.INT32, true, false);
        Property prop2 = TestDataFactory.createProperty("id", DataType.INT32, true, false);
        Property prop3 = TestDataFactory.createProperty("name", DataType.STRING, false, true);

        // Note: Property class doesn't override equals(), so this tests object identity
        Assert.assertNotEquals(prop1, prop2); // Different objects
        Assert.assertNotEquals(prop1, prop3); // Different properties

        // Same object reference
        Property sameRef = prop1;
        Assert.assertEquals(prop1, sameRef);
    }

    @Test
    public void testPropertyImmutability() {
        Property prop = TestDataFactory.createProperty("test", DataType.INT32, true, false);

        // Properties should be immutable (final fields)
        TestVerificationUtils.verifyProperty(prop, "test", true, false);
        Assert.assertEquals(DataType.INT32, prop.getDataType());

        // These values should not change after construction
        TestVerificationUtils.verifyProperty(prop, "test", true, false);
        Assert.assertEquals(DataType.INT32, prop.getDataType());
    }
}

File: go/graphar/info/property_test.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package info

import (
	"errors"
	"testing"

	"github.com/apache/incubator-graphar/go/graphar/types"
)

func TestNewPropertyDefaults(t *testing.T) {
	t.Parallel()
	// Primary: not nullable.
	p, err := NewProperty("id", types.Int64(), true)
	if err != nil {
		t.Fatalf("NewProperty: %v", err)
	}
	if p.Cardinality != types.CardinalitySingle {
		t.Errorf("default cardinality = %v, want single", p.Cardinality)
	}
	if p.IsNullable {
		t.Errorf("primary property default IsNullable should be false")
	}
	// Non-primary: nullable by default, matching the loader default.
	np, err := NewProperty("name", types.String(), false)
	if err != nil {
		t.Fatalf("NewProperty: %v", err)
	}
	if !np.IsNullable {
		t.Errorf("non-primary property default IsNullable should be true")
	}
}

func TestPropertyValidateRejectsEmptyName(t *testing.T) {
	t.Parallel()
	p := Property{Type: types.Int64()}
	err := p.Validate()
	if !errors.Is(err, ErrInvalidProperty) {
		t.Fatalf("expected ErrInvalidProperty, got %v", err)
	}
	var ve *ValidationError
	if !errors.As(err, &ve) || ve.Field != "name" {
		t.Errorf("expected field=name, got %+v", ve)
	}
}

func TestPropertyValidatePrimaryNotNullable(t *testing.T) {
	t.Parallel()
	p := Property{Name: "id", Type: types.Int64(), IsPrimary: true, IsNullable: true}
	err := p.Validate()
	if !errors.Is(err, ErrInvalidProperty) {
		t.Fatalf("expected ErrInvalidProperty, got %v", err)
	}
}

func TestPropertyEqual(t *testing.T) {
	t.Parallel()
	a := Property{Name: "x", Type: types.Int32(), Cardinality: types.CardinalitySingle}
	if !a.Equal(a) {
		t.Error("reflexive equality failed")
	}
	cases := []struct {
		mutate func(p *Property)
		label  string
	}{
		{func(p *Property) { p.Name = "y" }, "name"},
		{func(p *Property) { p.Type = types.Int64() }, "type"},
		{func(p *Property) { p.IsPrimary = true }, "primary"},
		{func(p *Property) { p.IsNullable = true }, "nullable"},
		{func(p *Property) { p.Cardinality = types.CardinalityList }, "cardinality"},
	}
	for _, c := range cases {
		b := a
		c.mutate(&b)
		if a.Equal(b) {
			t.Errorf("equal returned true after mutating %s", c.label)
		}
	}
}

func TestNewPropertyValidatesEmptyName(t *testing.T) {
	t.Parallel()
	_, err := NewProperty("", types.Int64(), false)
	if !errors.Is(err, ErrInvalidProperty) {
		t.Fatalf("expected ErrInvalidProperty, got %v", err)
	}
}

func TestPropertyValidateRejectsZeroType(t *testing.T) {
	t.Parallel()
	// A hand-constructed Property whose Type was never set must be rejected
	// rather than silently serialising as "bool".
	p := Property{Name: "id", IsPrimary: true}
	err := p.Validate()
	if !errors.Is(err, ErrInvalidProperty) {
		t.Fatalf("expected ErrInvalidProperty for unset type, got %v", err)
	}
}

func TestNewPropertyRejectsMalformedType(t *testing.T) {
	t.Parallel()
	bad := []struct {
		name string
		dt   types.DataType
	}{
		{"list of unset", types.List(types.DataType{})},
		{"list of bool", types.List(types.Bool())},
		{"list of date", types.List(types.Date())},
		{"nested list", types.List(types.List(types.Int32()))},
		{"empty user-defined", types.UserDefined("")},
	}
	for _, c := range bad {
		if _, err := NewProperty("x", c.dt, false); !errors.Is(err, ErrInvalidProperty) {
			t.Errorf("NewProperty(%s) = %v, want ErrInvalidProperty", c.name, err)
		}
	}
}

func TestNewPropertyAcceptsSupportedTypes(t *testing.T) {
	t.Parallel()
	ok := []types.DataType{
		types.List(types.Int32()), types.List(types.String()),
		types.UserDefined("geometry"),
	}
	for _, dt := range ok {
		if _, err := NewProperty("x", dt, false); err != nil {
			t.Errorf("NewProperty(%s) = %v, want nil", dt, err)
		}
	}
}

// TestPropertyValidateErrorChain guards that a bad data type surfaces both the
// property sentinel and the underlying data-type sentinel via errors.Is.
func TestPropertyValidateErrorChain(t *testing.T) {
	t.Parallel()
	p := Property{Name: "geo", Type: types.UserDefined("string")} // collides with builtin
	err := p.Validate()
	if err == nil {
		t.Fatal("expected error for colliding user-defined type")
	}
	if !errors.Is(err, ErrInvalidProperty) {
		t.Errorf("missing ErrInvalidProperty in chain: %v", err)
	}
	if !errors.Is(err, types.ErrInvalidDataType) {
		t.Errorf("missing types.ErrInvalidDataType in chain: %v", err)
	}
}
